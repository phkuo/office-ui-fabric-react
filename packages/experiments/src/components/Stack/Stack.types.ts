import { IStyle } from '../../Styling';
import { IStatelessComponent, IStyleableComponentProps } from '../../Foundation';

export type Alignment = 'start' | 'end' | 'center' | 'space-between' | 'space-around' | 'space-evenly' | 'baseline' | 'stretch';

export type Omit<T, K extends keyof T> = Pick<T, Exclude<keyof T, K>>;

export type IStackComponent = IStatelessComponent<IStackProps, IStackStyles>;

// contains the members of IStackProps that are common to both VerticalStack and HorizontalStack
export type IPartialStackProps = Omit<
  IStackProps,
  'verticalAlignment' | 'horizontalAlignment' | 'horizontal' | 'verticalGap' | 'horizontalGap' | 'wrap' | 'styles'
>;

export interface IStackProps extends IStyleableComponentProps<IStackProps, IStackStyles>, React.HTMLAttributes<HTMLElement> {
  /**
   * How to render the Stack.
   */
  as?: string | React.ReactType<IPartialStackProps>;

  /**
   * CSS class name used to style the Stack.
   */
  className?: string;

  /**
   * Inline styling.
   */
  style?: React.CSSProperties;

  /**
   * Whether to render Stack child elements horizontally.
   */
  horizontal?: boolean;

  /**
   * How to align Stack child elements horizontally (along the x-axis).
   */
  horizontalAlignment?: Alignment;

  /**
   * How to align Stack child elements vertically (along the y-axis).
   */
  verticalAlignment?: Alignment;

  /**
   * Whether the Stack should take up 100% of the width of its parent.
   */
  fillHorizontal?: boolean;

  /**
   * Whether the Stack should take up 100% of the height of its parent.
   */
  fillVertical?: boolean;

  /**
   * Whether Stack child elements should shrink to fit the available space.
   */
  shrinkItems?: boolean;

  /**
   * How much to grow the Stack in proportion to its siblings.
   */
  grow?: boolean | number | 'inherit' | 'initial' | 'unset';

  /**
   * Spacing between Stack child elements.
   */
  gap?: number | string;

  /**
   * Maximum width of the Stack.
   */
  maxWidth?: number | string;

  /**
   * Maximum height of the Stack.
   */
  maxHeight?: number | string;

  /**
   * Padding of the Stack.
   */
  padding?: number | string;

  /**
   * Whether Stack children should wrap onto multiple rows (HorizontalStacks only).
   */
  wrap?: boolean;

  /**
   * Horizontal gap between Stack child elements.
   */
  horizontalGap?: number | string;

  /**
   * Vertical gap between Stack child elements.
   */
  verticalGap?: number | string;
}

export interface IStackStyles {
  /**
   * Style set for the root element.
   */
  root: IStyle;
}
