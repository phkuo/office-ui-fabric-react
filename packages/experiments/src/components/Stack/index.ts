export * from './StackItem/StackItem';
export * from './VerticalStack/VerticalStack';
export * from './VerticalStack/VerticalStack.types';
export * from './HorizontalStack/HorizontalStack';
export * from './HorizontalStack/HorizontalStack.types';
