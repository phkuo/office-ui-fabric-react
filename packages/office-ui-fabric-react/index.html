<!doctype html>
<html>

<head>
  <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0, maximum-scale=1.0" user-scalable="no" />

  <title>Office UI Fabric React Examples</title>

</head>

<body>
  <div id="content"></div>

  <script type="text/javascript">
    var isProduction = (location.pathname === '/fabric-react/master/');

    function loadScript(n) {
      for (var i = 0; i < n.length; i++) {
        var s = document.createElement('script');
        s.src = n[i];
        s.async = false;
        document.body.appendChild(s);
      }
    }
    loadScript(isProduction ? [
      '//cdnjs.cloudflare.com/ajax/libs/react/15.4.0/react.min.js',
      '//cdnjs.cloudflare.com/ajax/libs/react/15.4.0/react-dom.min.js',
      'dist/demo-vendor.min.js',
      'dist/demo-components.min.js',
      'dist/demo-app.min.js'
    ] : [
      '//cdnjs.cloudflare.com/ajax/libs/react/15.4.0/react.js',
      '//cdnjs.cloudflare.com/ajax/libs/react/15.4.0/react-dom.js',
      'dist/demo-vendor.js',
      'dist/demo-components.js',
      'dist/demo-app.js'
    ]);
    </script>
</body>

</html>