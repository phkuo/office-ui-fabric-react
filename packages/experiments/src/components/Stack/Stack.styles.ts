import { IStackComponent, IStackStyles } from './Stack.types';
import { parseGap, parsePadding } from './StackUtils';

const nameMap: { [key: string]: string } = {
  start: 'flex-start',
  end: 'flex-end'
};

export const styles: IStackComponent['styles'] = props => {
  const {
    fillHorizontal,
    fillVertical,
    maxWidth,
    maxHeight,
    horizontal,
    grow,
    wrap,
    padding,
    horizontalAlignment,
    verticalAlignment,
    horizontalGap,
    verticalGap,
    shrinkItems,
    theme,
    className
  } = props;

  const hGap = parseGap(horizontalGap, theme);
  const vGap = parseGap(verticalGap, theme);

  // styles to be applied to all direct children regardless of wrap or direction
  const childStyles = {
    whiteSpace: 'nowrap',
    textOverflow: 'ellipsis'
  };

  // selectors to be applied regardless of wrap or direction
  const commonSelectors = {
    '> *:empty': {
      display: 'none'
    },

    // flexShrink styles are applied by the StackItem
    '> *:not(.ms-StackItem)': {
      flexShrink: shrinkItems ? 1 : 0
    }
  };

  return {
    root: [
      theme.fonts.medium,
      {
        display: 'flex',
        flexDirection: horizontal ? 'row' : 'column',
        flexWrap: 'nowrap',
        width: fillHorizontal && !wrap ? '100%' : 'auto',
        height: fillVertical && !wrap ? '100%' : 'auto',
        maxWidth,
        maxHeight,
        padding: parsePadding(padding, theme),
        boxSizing: 'border-box'
      },
      grow && {
        flexGrow: grow === true ? 1 : grow,
        overflow: 'hidden'
      },
      horizontalAlignment && {
        [horizontal ? 'justifyContent' : 'alignItems']: nameMap[horizontalAlignment] || horizontalAlignment
      },
      verticalAlignment && {
        [horizontal ? 'alignItems' : 'justifyContent']: nameMap[verticalAlignment] || verticalAlignment
      },
      wrap && {
        selectors: {
          '> *': {
            margin: `${0.5 * vGap.value}${vGap.unit} ${0.5 * hGap.value}${hGap.unit}`,

            // avoid unnecessary calc() calls if horizontal gap is 0
            maxWidth: hGap.value === 0 ? '100%' : `calc(100% - ${hGap.value}${hGap.unit})`,

            ...childStyles
          },
          ...commonSelectors
        }
      },
      !wrap && {
        selectors: {
          '> *': childStyles,

          // apply gap margin to every direct child except the first non-empty direct child
          '> *:not(:empty) ~ *': [
            horizontal && {
              marginLeft: `${hGap.value}${hGap.unit}`
            },
            !horizontal && {
              marginTop: `${vGap.value}${vGap.unit}`
            }
          ],
          ...commonSelectors
        }
      },
      className
    ]
    // TODO: this cast may be hiding some potential issues with styling and name
    //        lookups and should be removed
  } as IStackStyles;
};
