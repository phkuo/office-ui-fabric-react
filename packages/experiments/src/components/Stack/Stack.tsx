import * as React from 'react';
import { createStatelessComponent, IStyleableComponentProps } from '../../Foundation';
import StackItem from './StackItem/StackItem';
import { IStackItemProps, IStackItemStyles } from './StackItem/StackItem.types';
import { IStackComponent, IStackProps, IStackStyles } from './Stack.types';
import { styles } from './Stack.styles';
import { mergeStyles } from '../../Styling';
import { getNativeProps, htmlElementProperties } from '../../Utilities';

const StackItemType = (<StackItem /> as React.ReactElement<IStackItemProps> & IStyleableComponentProps<IStackItemProps, IStackItemStyles>)
  .type;

const view: IStackComponent['view'] = props => {
  const { as: RootType = 'div', classNames, shrinkItems, wrap, ...rest } = props;

  const stackChildren: (React.ReactChild | null)[] = React.Children.map(
    props.children,
    (child: React.ReactElement<IStackItemProps>, index: number) => {
      if (!child) {
        return null;
      }

      const defaultItemProps: IStackItemProps = {
        shrink: shrinkItems,
        className: child.props ? child.props.className : undefined
      };

      if (child.type === StackItemType) {
        // If child is a StackItem, we need to pass down the className of ITS first child to the StackItem for mergeStylesSet to work
        const children = child.props ? child.props.children : undefined;
        const stackItemFirstChildren = React.Children.toArray(children) as React.ReactElement<{ className?: string }>[];
        const stackItemFirstChild = stackItemFirstChildren && stackItemFirstChildren[0];

        // pass down both the className on the StackItem as well as the className on its first child
        let mergedClassName = defaultItemProps.className;
        if (stackItemFirstChild && stackItemFirstChild.props && stackItemFirstChild.props.className) {
          mergedClassName = mergeStyles(mergedClassName, stackItemFirstChild.props.className);
        }

        return React.cloneElement(child, {
          ...defaultItemProps,
          ...child.props,
          className: mergedClassName
        });
      }

      return child;
    }
  );

  const nativeProps = getNativeProps(rest, htmlElementProperties);

  return (
    <RootType {...nativeProps} className={classNames.root}>
      {stackChildren}
    </RootType>
  );
};

const StackStatics = {
  Item: StackItem,
  defaultProps: {}
};
type IStackStatics = typeof StackStatics;

export const Stack: React.StatelessComponent<IStackProps> & {
  Item: React.StatelessComponent<IStackItemProps>;
} = createStatelessComponent<IStackProps, IStackStyles, IStackStatics>({
  displayName: 'Stack',
  styles,
  view,
  statics: StackStatics
});

export default Stack;
