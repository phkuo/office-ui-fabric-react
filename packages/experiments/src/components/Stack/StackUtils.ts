/**
 * Functions used by Stack components to simplify style-related computations
 */

import { Alignment } from './Stack.types';
import { ITheme } from '../../Styling';

const horizontalAlignmentMap: { [key: string]: string } = {
  left: 'start',
  right: 'end'
};

const verticalAlignmentMap: { [key: string]: string } = {
  top: 'start',
  bottom: 'end'
};

// Converts the alignment expected by a HorizontalStack to an alignment expected by a Stack.
export const getHorizontalAlignment = (alignment: string | undefined): Alignment => {
  return (horizontalAlignmentMap[alignment!] || alignment) as Alignment;
};

// Converts the alignment expected by a VerticalStack to an alignment expected by a Stack.
export const getVerticalAlignment = (alignment: string | undefined): Alignment => {
  return (verticalAlignmentMap[alignment!] || alignment) as Alignment;
};

// Helper function that converts a themed spacing key (if given) to the corresponding themed spacing value.
const _getThemedSpacing = (space: string, theme: ITheme): string => {
  if (theme.spacing.hasOwnProperty(space)) {
    return theme.spacing[space as keyof typeof theme.spacing];
  }
  return space;
};

/**
 * Takes in a gap size in either a CSS-style format (e.g. 10 or "10px")
 *  or a key of a themed spacing value (e.g. "s1").
 * Returns the separate numerical value of the padding (e.g. 10)
 *  and the CSS unit (e.g. "px").
 */
export const parseGap = (gap: number | string | undefined, theme: ITheme): { value: number; unit: string } => {
  if (gap === undefined || gap === '') {
    return {
      value: 0,
      unit: 'px'
    };
  }

  if (typeof gap === 'number') {
    return {
      value: gap,
      unit: 'px'
    };
  }

  const stringGap = _getThemedSpacing(gap, theme);

  const numericalPart = parseFloat(stringGap);
  const numericalValue = isNaN(numericalPart) ? 0 : numericalPart;
  const numericalString = isNaN(numericalPart) ? '' : numericalPart.toString();

  const unitPart = stringGap.substring(numericalString.toString().length);

  return {
    value: numericalValue,
    unit: unitPart || 'px'
  };
};

/**
 * Takes in a padding in a CSS-style format (e.g. 10, "10px", "10px 10px", etc.)
 *  where the separate padding values can also be the key of a themed spacing value
 *  (e.g. "s1 m", "10px l1 20px l2", etc.).
 * Returns a CSS-style padding.
 */
export const parsePadding = (padding: number | string | undefined, theme: ITheme): number | string | undefined => {
  if (padding === undefined || typeof padding === 'number' || padding === '') {
    return padding;
  }

  const paddingValues = padding.split(' ');
  if (paddingValues.length < 2) {
    return _getThemedSpacing(padding, theme);
  }

  return paddingValues.reduce((padding1: string, padding2: string) => {
    return _getThemedSpacing(padding1, theme) + ' ' + _getThemedSpacing(padding2, theme);
  });
};
