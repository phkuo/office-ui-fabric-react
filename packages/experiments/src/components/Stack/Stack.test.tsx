import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { mergeStyles } from 'office-ui-fabric-react';

import { Stack } from './Stack';

const sampleClass = mergeStyles({
  background: 'red'
});

describe('Stack', () => {
  it('renders vertical Stack correctly', () => {
    const component = renderer.create(
      <Stack>
        <div>Item 1</div>
        <div>Item 2</div>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders horizontal Stack correctly', () => {
    const component = renderer.create(
      <Stack horizontal>
        <div>Item 1</div>
        <div>Item 2</div>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Stack with vertical and horizontal alignment correctly', () => {
    const component = renderer.create(
      <Stack horizontalAlignment="start" verticalAlignment="end">
        <div>Item 1</div>
        <div>Item 2</div>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Stack with vertical and horizontal fill correctly', () => {
    const component = renderer.create(
      <Stack fillHorizontal fillVertical>
        <div>Item 1</div>
        <div>Item 2</div>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Stack with grow correctly', () => {
    const component = renderer.create(
      <Stack>
        <Stack grow={5}>Item 1</Stack>
        <Stack grow={2}>Item 2</Stack>
        <Stack grow>Item 3</Stack>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Stack with StackItems correctly', () => {
    const component = renderer.create(
      <Stack>
        <Stack.Item>Item 1</Stack.Item>
        <Stack.Item>Item 2</Stack.Item>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Stack with a gap correctly', () => {
    const component = renderer.create(
      <Stack gap={10}>
        <Stack.Item>Item 1</Stack.Item>
        <Stack.Item>Item 2</Stack.Item>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders Stack with shrinking StackItems correctly', () => {
    const component = renderer.create(
      <Stack shrinkItems>
        <Stack.Item>Item 1</Stack.Item>
        <Stack.Item>Item 2</Stack.Item>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('accepts custom className on child items', () => {
    const component = renderer.create(
      <Stack>
        <Stack.Item className={sampleClass}>Item 1</Stack.Item>
        <div className={sampleClass}>Item 2</div>
      </Stack>
    );
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
