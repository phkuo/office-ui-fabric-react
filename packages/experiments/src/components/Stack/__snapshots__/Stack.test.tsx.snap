// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stack accepts custom className on child items 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <span
    className=
        ms-StackItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          background: red;
          flex-shrink: 0;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          width: auto;
        }
  >
    Item 1
  </span>
  <div
    className=

        {
          background: red;
        }
  >
    Item 2
  </div>
</div>
`;

exports[`Stack renders Stack with StackItems correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <span
    className=
        ms-StackItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          flex-shrink: 0;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          width: auto;
        }
  >
    Item 1
  </span>
  <span
    className=
        ms-StackItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          flex-shrink: 0;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          width: auto;
        }
  >
    Item 2
  </span>
</div>
`;

exports[`Stack renders Stack with a gap correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <span
    className=
        ms-StackItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          flex-shrink: 0;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          width: auto;
        }
  >
    Item 1
  </span>
  <span
    className=
        ms-StackItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          flex-shrink: 0;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          width: auto;
        }
  >
    Item 2
  </span>
</div>
`;

exports[`Stack renders Stack with grow correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <div
    className=

        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          box-sizing: border-box;
          display: flex;
          flex-direction: column;
          flex-grow: 5;
          flex-wrap: nowrap;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          overflow: hidden;
          width: auto;
        }
        & > * {
          text-overflow: ellipsis;
          white-space: nowrap;
        }
        & > *:not(:empty) ~ * {
          margin-top: 0px;
        }
        & > *:empty {
          display: none;
        }
        & > *:not(.ms-StackItem) {
          flex-shrink: 0;
        }
  >
    Item 1
  </div>
  <div
    className=

        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          box-sizing: border-box;
          display: flex;
          flex-direction: column;
          flex-grow: 2;
          flex-wrap: nowrap;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          overflow: hidden;
          width: auto;
        }
        & > * {
          text-overflow: ellipsis;
          white-space: nowrap;
        }
        & > *:not(:empty) ~ * {
          margin-top: 0px;
        }
        & > *:empty {
          display: none;
        }
        & > *:not(.ms-StackItem) {
          flex-shrink: 0;
        }
  >
    Item 2
  </div>
  <div
    className=

        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          box-sizing: border-box;
          display: flex;
          flex-direction: column;
          flex-grow: 1;
          flex-wrap: nowrap;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          overflow: hidden;
          width: auto;
        }
        & > * {
          text-overflow: ellipsis;
          white-space: nowrap;
        }
        & > *:not(:empty) ~ * {
          margin-top: 0px;
        }
        & > *:empty {
          display: none;
        }
        & > *:not(.ms-StackItem) {
          flex-shrink: 0;
        }
  >
    Item 3
  </div>
</div>
`;

exports[`Stack renders Stack with shrinking StackItems correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 1;
      }
>
  <span
    className=
        ms-StackItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          flex-shrink: 1;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          width: auto;
        }
  >
    Item 1
  </span>
  <span
    className=
        ms-StackItem
        {
          -moz-osx-font-smoothing: grayscale;
          -webkit-font-smoothing: antialiased;
          flex-shrink: 1;
          font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
          font-size: 14px;
          font-weight: 400;
          height: auto;
          width: auto;
        }
  >
    Item 2
  </span>
</div>
`;

exports[`Stack renders Stack with vertical and horizontal alignment correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        align-items: flex-start;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        justify-content: flex-end;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
</div>
`;

exports[`Stack renders Stack with vertical and horizontal fill correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: 100%;
        width: 100%;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
</div>
`;

exports[`Stack renders horizontal Stack correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: row;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-left: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
</div>
`;

exports[`Stack renders vertical Stack correctly 1`] = `
<div
  className=

      {
        -moz-osx-font-smoothing: grayscale;
        -webkit-font-smoothing: antialiased;
        box-sizing: border-box;
        display: flex;
        flex-direction: column;
        flex-wrap: nowrap;
        font-family: 'Segoe UI', 'Segoe UI Web (West European)', 'Segoe UI', -apple-system, BlinkMacSystemFont, 'Roboto', 'Helvetica Neue', sans-serif;
        font-size: 14px;
        font-weight: 400;
        height: auto;
        width: auto;
      }
      & > * {
        text-overflow: ellipsis;
        white-space: nowrap;
      }
      & > *:not(:empty) ~ * {
        margin-top: 0px;
      }
      & > *:empty {
        display: none;
      }
      & > *:not(.ms-StackItem) {
        flex-shrink: 0;
      }
>
  <div>
    Item 1
  </div>
  <div>
    Item 2
  </div>
</div>
`;
